File: 1022-unique-paths-iii/unique-paths-iii.java
class Solution {
    int result = 0, row = 0, col = 0, coverage = 0;
    int directions[][] = {{1,0}, {0,1}, {-1,0}, {0, -1}};

    void dfs(int r, int c, int curr, int[][] grid){
        if (grid[r][c] == -1){
            return;
        }

        if (grid[r][c] == 2){
            if(curr == coverage){
                result++;
            }
            return;
        }

        for (int[] direction: directions)
        {
            int x = direction[0] + r;
            int y = direction[1] + c;
            if (0 > x || 0 > y || x == row || y == col || grid[x][y] == -1)
                continue;
            
            grid[r][c] = -1;
            dfs(x, y, (grid[x][y] == 0 ? curr + 1 : curr), grid);
            grid[r][c] = 0;
        }
    }

    public int uniquePathsIII(int[][] grid) {
        row = grid.length;
        col = grid[0].length;
        int start[] = new int[2]; 
        coverage = 0;
        for (int r = 0; r < row; r++){
            for (int c = 0; c < col; c++){
                if (grid[r][c] == 1){
                    start[0] = r;
                    start[1] = c;
                }
                else if (grid[r][c] == 0){
                    coverage++;
                }
            }
        }

        dfs(start[0], start[1], 0, grid);
        return result;
    }
}

File: 1022-unique-paths-iii/unique-paths-iii.py
class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        row = len(grid)
        col = len(grid[0])
        result = 0
        start = None
        cover = 0
        for i in range(row):
            for j in range(col):
                if grid[i][j] == 1:
                    start = (i, j)
                elif grid[i][j] == 0:
                    cover += 1

        def dfs(x, y, curr):
            nonlocal result
            if grid[x][y] == -1:
                return
            if grid[x][y] == 2:
                if curr == cover:
                    result += 1
                return 
            for i, j in [[1, 0], [0, 1], [-1, 0], [0, -1]]:
                ii, jj = i + x, j + y
                
                if 0 > ii or ii == row or 0 > jj or jj == col or grid[ii][jj] == -1:
                    continue
                
                grid[x][y] = -1
                dfs(ii, jj, (curr + 1) if grid[ii][jj] == 0 else curr)
                grid[x][y] = 0

        dfs(start[0], start[1], 0)
        return result
